### PR TITLE
[backport#19601] Add single sha256 call to CHashWriter

### PR DESCRIPTION
**Pull request by EyeOfPython**
Part 2 of 3.

### CODE REVIEW

File: src/hash.h [hunk 1]
```diff
@@ -111,23 +111,38 @@ class CHashWriter {
     int GetVersion() const { return nVersion; }
 
     void write(const char *pch, size_t size) {
-        ctx.Write({(const uint8_t *)pch, size});
+        ctx.Write((const uint8_t *)pch, size);
     }
 
-    // invalidates the object
+    /** Compute the double-SHA256 hash of all data written to this object.
```
Comment (schancel) [inline src/hash.h:117]: How is OP_CDS doing the single round of SHA256 right now?

Comment (schancel) [reply src/hash.h:117]:
```
                        if (vchSig.size()) {
                            valtype vchHash(32);
                            CSHA256()
                                .Write(vchMessage.data(), vchMessage.size())
                                .Finalize(vchHash.data());
                            fSuccess = checker.VerifySignature(
                                vchSig, CPubKey(vchPubKey), uint256(vchHash));
                            metrics.nSigChecks += 1;
```

May want to refactor this as you're doing this. IDK.

Comment (EyeOfPython) [reply src/hash.h:117]: Great idea.

File: src/hash.h [hunk 2]
```diff
@@ -112,23 +112,38 @@ class CHashWriter {
     int GetVersion() const { return nVersion; }
 
     void write(const char *pch, size_t size) {
-        ctx.Write({(const uint8_t *)pch, size});
+        ctx.Write((const uint8_t *)pch, size);
```
Comment (schancel) [inline src/hash.h:115]: Why does this need to change? Seems like an optimization of some sort?

File: src/hash.h [hunk 3]
```diff
@@ -112,23 +112,38 @@ class CHashWriter {
...
 
-    // invalidates the object
+    /** Compute the double-SHA256 hash of all data written to this object.
+     *
+     * Invalidates this object.
+     */
     uint256 GetHash() {
         uint256 result;
-        ctx.Finalize(result);
+        ctx.Finalize(result.begin());
```
Comment (schancel) [inline src/hash.h:124]: Why does this need to change? Is it just to reset it?

**Review by schancel — APPROVED** [review]

File: src/hash.h (hunk 2 above)
Comment (EyeOfPython) [reply src/hash.h:115]: No, they just have different signatures:

```
CHash256 &Write(Span<const uint8_t> input);
vs
CSHA256 &Write(const uint8_t *data, size_t len);
```

🤷‍♂️

File: src/hash.h (hunk 3 above)
Comment (EyeOfPython) [reply src/hash.h:124]: Again, different signatures:

```
void CHash256::Finalize(Span<uint8_t> output);
vs.
void CSHA256::Finalize(uint8_t hash[32]);
```

Should probably be refactored at some point, but not yet.

**Pull request merged**